### PR TITLE
Data slice mutation support

### PR DESCRIPTION
**Pull request by phausler**
Mutation of Data slices had issues when it comes to re-indexing to the base slice region. Previously when a slice of Data was mutated the re-calculation of the slice region was incorrectly assigned to a relative region to the previous backing. Now the slice region will copy the specific region out creating a new data upon CoW semantics and offset the index region upon access. Mutating a non CoW slice will retain the previous region backing but adjust the remaining allocation buffer accordingly. To validate this I have added tests for combinations that should approximate all possible combinations of mutations, backing storages and CoW scenarios.

This fixes the following issues:
rdar://problem/34206043
SR-5887
SR-5873
SR-5810

tl;dr – Slicing a CoW should result in fillet mignon not ground chuck.

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci please test

**Comment by phausler** [comment]
@swift-ci please test

File: stdlib/public/SDK/Foundation/Data.swift [hunk 1]
```diff
@@ -98,47 +102,140 @@ public final class _DataStorage {
     public var _needToZero: Bool
     public var _deallocator: ((UnsafeMutableRawPointer, Int) -> Void)?
     public var _backing: Backing = .swift
+    public var _offset: Int
     
     public var bytes: UnsafeRawPointer? {
         @inline(__always)
         get {
             switch _backing {
             case .swift:
```
Comment (itaiferber) [inline stdlib/public/SDK/Foundation/Data.swift]: Don't know how much this matters, but you should be able to shorten this a little bit:

```swift
let bytesPtr: UnsafeRawPointer?
switch _backing {
case .swift: fallthrough
case .immutable: fallthrough
case .mutable: bytesPtr = _bytes

case .customReference(let d): bytesPtr = d.bytes
case .customMutableReference(let d): bytesPtr = d.bytes
}

return bytesPtr?.advanced(by: -_offset)
```

**Review by itaiferber — APPROVED** [review]
Looks good to me overall — minor code style comments, but nothing blocking. Glad we have such a thorough list of slicing tests now!

File: stdlib/public/SDK/Foundation/Data.swift [hunk 2]
```diff
@@ -98,47 +102,140 @@ public final class _DataStorage {
...
-                return UnsafeRawPointer(_bytes)
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: probably can be; for now I was avoiding doing massive refactoring with the un-related switches to reduce risk.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 3]
```diff
@@ -98,47 +102,140 @@ public final class _DataStorage {
...
+
+    @discardableResult
+    public func withUnsafeBytes<Result>(in range: Range<Int>, apply: (UnsafeRawBufferPointer) throws -> Result) rethrows -> Result {
+        switch _backing {
+        case .swift: fallthrough
+        case .immutable: fallthrough
+        case .mutable:
+            return try apply(UnsafeRawBufferPointer(start: _bytes?.advanced(by: range.lowerBound - _offset), count: Swift.min(range.count, _length)))
+        case .customReference(let d):
+            if d._isCompact() {
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: again, I think this can be done in a subsequent change

File: stdlib/public/SDK/Foundation/Data.swift [hunk 4]
```diff
@@ -176,20 +273,18 @@ public final class _DataStorage {
         }
     }
     
-    public func enumerateBytes(_ block: (_ buffer: UnsafeBufferPointer<UInt8>, _ byteIndex: Data.Index, _ stop: inout Bool) -> Void) {
+    public func enumerateBytes(in range: Range<Int>, _ block: (_ buffer: UnsafeBufferPointer<UInt8>, _ byteIndex: Data.Index, _ stop: inout Bool) -> Void) {
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: not fully certain that matters just yet, but definitely once ABI stability hits we should consider these types of changes; thankfully the instance of _DataStorage itself is versioned so we should be in the clear here.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 5]
```diff
@@ -349,6 +444,8 @@ public final class _DataStorage {
         
     }
     
+    ////////////***** AUDITED UP TO HERE
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: oops, will remove that. it was my first pass on the offset additions and I forgot to remove it when I did the remainder of the audits

File: stdlib/public/SDK/Foundation/Data.swift [hunk 6]
```diff
@@ -1137,7 +1273,7 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
         }
         @inline(__always)
         set {
-            precondition(count >= 0, "Count must be positive")
+            precondition(count >= 0, "count must be positive")
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: not sure why this actually changed, perhaps it was the sync across sclf?

File: stdlib/public/SDK/Foundation/Data.swift [hunk 7]
```diff
@@ -1329,22 +1462,20 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
...
-        count += estimatedCount
-        for byte in newElements {
-            let newIndex = idx + 1
-            if newIndex > count {
-                count = newIndex
-            }
-            self[idx] = byte
-            idx = newIndex
+        var ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: estimatedCount)
+        defer { ptr.deallocate(capacity: estimatedCount) }
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: yes! I think that would be a reasonable change to avoid perf regressions here

File: stdlib/public/SDK/Foundation/Data.swift [hunk 8]
```diff
@@ -1329,22 +1462,20 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
...
+        let buffer = UnsafeMutableBufferPointer(start: ptr, count: estimatedCount)
+        var (iterator, endPoint)  = newElements._copyContents(initializing: buffer)
```
[the remark this answers is not in the document]
Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: not fully certain that can be done easily without a more extensive refactor

**Comment by phausler** [comment]
As a note: this patch is intended to be as minimal as possible to reduce the risk of regressions since we are considering this for an earlier submission. So some improvements might make it in with later work we have scheduled for integrating other functionality.

**Comment by michael-lehew** [comment]
> tl;dr – Slicing a CoW should result in fillet mignon not ground chuck.
This is a winning comment regardless of the rest of the content (reading now) :)

File: stdlib/public/SwiftShims/FoundationOverlayShims.h [hunk 9]
```diff
@@ -29,3 +32,39 @@
...
+    size_t capacity;
+    _Bool onStack;
+} _ConditionalAllocationBuffer;
+
+static inline _Bool _resizeConditionalAllocationBuffer(_ConditionalAllocationBuffer *_Nonnull buffer, size_t amt) {
+    size_t amount = malloc_good_size(amt);
+    if (amount <= buffer->capacity) { return true; }
+    void *newMemory;
+    if (buffer->onStack) {
+        newMemory = malloc(amount);
```
Comment (michael-lehew) [inline stdlib/public/SwiftShims/FoundationOverlayShims.h]: check for malloc fail and abort?

File: stdlib/public/SwiftShims/FoundationOverlayShims.h [hunk 10]
```diff
@@ -29,3 +32,39 @@
...
+        memcpy(newMemory, buffer->memory, buffer->capacity);
+        buffer->onStack = false;
+    } else {
+        newMemory = realloc(buffer->memory, amount);
```
Comment (michael-lehew) [inline stdlib/public/SwiftShims/FoundationOverlayShims.h]: this will leak if realloc fails (the normal realloc nasty bug)

**Review by michael-lehew — APPROVED** [review]
This looks good to me overall, the tests are great improvement.

**Comment by phausler** [comment]
@swift-ci please test

File: stdlib/public/SDK/Foundation/Data.swift [hunk 11]
```diff
@@ -198,7 +293,7 @@ public final class _DataStorage {
             d.enumerateBytes { (ptr, range, stop) in
                 var stopv = false
                 let bytePtr = ptr.bindMemory(to: UInt8.self, capacity: range.length)
-                block(UnsafeBufferPointer(start: bytePtr, count: range.length), range.location, &stopv)
+                block(UnsafeBufferPointer(start: bytePtr, count: range.length), range.location - _offset, &stopv)
```
Comment (kperryua) [inline stdlib/public/SDK/Foundation/Data.swift]: As I mentioned to you separately, these still need to be clamped to the `range` parameter passed to `enumerateBytes`. It might be a good idea to avoid the variable aliasing as well.

Comment (phausler) [reply stdlib/public/SDK/Foundation/Data.swift]: yep, working on a test and patch for that one; thanks again

File: stdlib/public/SDK/Foundation/Data.swift [hunk 12]
```diff
@@ -398,14 +493,51 @@ public final class _DataStorage {
...
+        case .swift: fallthrough
+        case .immutable: fallthrough
+        case .mutable:
+            return _bytes!.advanced(by: index - _offset).assumingMemoryBound(to: UInt8.self).pointee
+        case .customReference(let d):
+            if d._isCompact() {
+                return d.bytes.advanced(by: index - _offset).assumingMemoryBound(to: UInt8.self).pointee
+            } else {
+                var byte: UInt8 = 0
+                d.enumerateBytes { (ptr, range, stop) in
```
Comment (kperryua) [inline stdlib/public/SDK/Foundation/Data.swift]: It might be more code-size-efficient to just call `getBytes:range:NSMakeRange(offsetIndex, 1)` here. Discontiguous data implementations of that method should do essentially the same thing that you have written here.

File: stdlib/public/SDK/Foundation/Data.swift [hunk 13]
```diff
@@ -1329,22 +1460,26 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
         if !isKnownUniquelyReferenced(&_backing) {
             _backing = _backing.mutableCopy(_sliceRange)
         }
-        _backing.append(buffer.baseAddress!, length: buffer.count * MemoryLayout<SourceType>.stride)
+        _backing.replaceBytes(in: NSRange(location: _sliceRange.upperBound, length: _backing.length - (_sliceRange.upperBound - _backing._offset)), with: buffer.baseAddress, length: buffer.count * MemoryLayout<SourceType>.stride)
```
Comment (kperryua) [inline stdlib/public/SDK/Foundation/Data.swift]: This seems odd to me from a principled/conceptual standpoint. It seems like the Data should not reaching outside of its `_sliceRange`, and instead that this should be `_backing.replaceBytes(in: NSRange(location: _sliceRange.upperBound, length: 0)…`, which is exactly how you'd implement an abstract "append" if "replace" was the only primitive available.

In practice, I get it. Either:
1. The slice is the last reference of a previously larger backing, and we don't need to keep any bytes after our `_sliceRange` around anymore.
or
2. `_backing` is newly allocated due to CoW and the length parameter ends up 0 anyway.

I feel vaguely like a slightly different abstraction would satisfy both the principled and practiced side, but I'm not sure what that'd be right now.

Do we ever have occasion to clean up the bytes that *precede* the slice range when the slice is the last reference?

File: stdlib/public/SwiftShims/FoundationOverlayShims.h [hunk 14]
```diff
@@ -29,3 +32,41 @@
...
+    }
+    if (newMemory == NULL) { return false; }
+    buffer->memory = newMemory;
+    buffer->capacity = amount;
+    return true;
+}
+
+static inline _Bool _withStackOrHeapBuffer(size_t amount, void (__attribute__((noescape)) ^ _Nonnull applier)(_ConditionalAllocationBuffer *_Nonnull)) {
+    _ConditionalAllocationBuffer buffer;
+    buffer.capacity = malloc_good_size(amount);
```
Comment (kperryua) [inline stdlib/public/SwiftShims/FoundationOverlayShims.h]: Why round up prior to determining `onStack`? If the requested amount of memory is 500 bytes, `malloc_good_size` will probably return 512 or higher, which puts it out of stack viability (on the main thread). From a quick reading, it seems like we intend for allocations of that size to be on the stack.

I'm curious what the benefit to `malloc_good_size` is here at all. I suppose if you expected it to grow, you'd want to `malloc_good_size` to give you some more time before needing to realloc, but I don't actually see `_resizeConditionalAllocationBuffer` being called anywhere.

File: test/stdlib/TestData.swift [hunk 15]
```diff
@@ -1215,74 +1287,2669 @@ class TestData : TestDataSuper {
...
+        data.withUnsafeMutableBytes { (ptr: UnsafeMutablePointer<UInt8>) in
+            ptr.advanced(by: 5).pointee = 0xFF
+        }
+        expectEqual(data, Data(bytes: [0, 1, 2, 3, 4, 0xFF, 6, 7, 8, 9]))
+    }
+    
+    func test_validateMutation_appendBytes() {
+        var data = Data(bytes: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+        data.append("hello", count: 5)
+        expectEqual(data[data.startIndex.advanced(by: 5)], 0x5)
```
Comment (kperryua) [inline test/stdlib/TestData.swift]: This test seems odd since it doesn't test that it actually appended the new bytes.

File: test/stdlib/TestData.swift [hunk 16]
```diff
@@ -1215,74 +1287,2669 @@ class TestData : TestDataSuper {
...
+    }
+    
+    func test_validateMutation_appendData() {
+        var data = Data(bytes: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+        let other = Data(bytes: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+        data.append(other)
+        expectEqual(data[data.startIndex.advanced(by: 9)], 9)
+        expectEqual(data[data.startIndex.advanced(by: 10)], 0)
```
Comment (kperryua) [inline test/stdlib/TestData.swift]: Should we test the length too?

Comment (phausler) [reply test/stdlib/TestData.swift]: probably I should change it to the full comparison version

File: test/stdlib/TestData.swift [hunk 17]
```diff
@@ -1215,74 +1287,2669 @@ class TestData : TestDataSuper {
...
+        holdReference(data) {
+            let replacement: [UInt8] = [0xFF, 0xFF]
+            let range: Range<Data.Index> = data.startIndex.advanced(by: 4)..<data.startIndex.advanced(by: 9)
+            data.replaceSubrange(range, with: replacement)
+            expectEqual(data, Data(bytes: [0x68, 0x65, 0x6c, 0x6c, 0xff, 0xff, 0x6c, 0x64, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06]))
+        }
+    }
+    
+    func test_validateMutation_cow_mutableBacking_replaceSubrangeWithBytes() {
+        var data = NSData(b
```
Comment (kperryua) [inline test/stdlib/TestData.swift]: Huh… I thought extending an AllOnesData's count filled it with ones. Guess I'm wrong?

Comment (kperryua) [inline test/stdlib/TestData.swift]: If `data` really is expected to be `[1, 0, 0, 0, 0, 0, 0, 0, 0, 0]` after `data.count = 10`, then this doesn't really test for anything.

**Comment by phausler** [comment]
@kperryua the latest change should address the enumeration of slice regions such that a slice of discontiguous data will enumerate regions that are restricted to the slice range.

**Comment by phausler** [comment]
@swift-ci please test

**Comment by CharlesJS** [comment]
> tl;dr – Slicing a CoW should result in fillet mignon not ground chuck.

Dang it, that's better than both my "The CoW got tipped" and "mutates like nuclear fallout" puns. Nicely done.
